sdk/java-sdk/src/main/scala/kalix/javasdk/impl/GrpcClients.scala

```diff
@@ -71,37 +74,67 @@ final class GrpcClients(system: ExtendedActorSystem) extends Extension {
       }
       .map(_ => Done))
 
-  def setSelfServiceName(deployedName: String): Unit = {
-    log.debug("Setting proxy name to: [{}]", deployedName)
-    selfServiceName = Some(deployedName)
+  def setProxyHostname(hostname: String): Unit = {
+    log.debug("Setting proxy hostname to: [{}]", hostname)
+    proxyHostname = Some(hostname)
+  }
+
+  def setIdentificationInfo(info: Option[IdentificationInfo]): Unit = {
+    log.debug("Setting identification info to name to: [{}]", info)
+    identificationInfo = info
   }
 
   def setSelfServicePort(port: Int): Unit = {
     log.debug("Setting port to: [{}]", port)
     selfPort = Some(port)
   }
 
-  def getComponentGrpcClient[T](serviceClass: Class[T]): T =
+  def getComponentGrpcClient[T](serviceClass: Class[T]): T = {
     getLocalGrpcClient(serviceClass)
+  }
 
-  def getProxyGrpcClient[T](serviceClass: Class[T]): T =
+  def getProxyGrpcClient[T](serviceClass: Class[T]): T = {
     getLocalGrpcClient(serviceClass)
+  }
 
+  /** This gets called from the action context to get a client to another service, and hence needs to add a
+   * service identification header (in dev/test mode) to ensure calls get associated with this service. */
   def getGrpcClient[T](serviceClass: Class[T], service: String): T =
-    getGrpcClient(serviceClass, service, port = 80)
+    getGrpcClient(serviceClass, service, port = 80, remoteAddHeader)
 
   /** Local gRPC clients point to services (user components or Kalix services) in the same deployable */
   private def getLocalGrpcClient[T](serviceClass: Class[T]): T = {
-    selfServiceName match {
-      case Some("localhost") => getGrpcClient(serviceClass, "localhost", selfPort.getOrElse(9000))
-      case Some(selfName)    => getGrpcClient(serviceClass, selfName)
+    proxyHostname match {
+      case Some("localhost") => getGrpcClient(serviceClass, "localhost", selfPort.getOrElse(9000), localAddHeader)
+      case Some(selfName)    => getGrpcClient(serviceClass, selfName, 80, localAddHeader)
```

Comment:
Not strictly related to this PR, more of a future improvement perhaps: if we pass proxy port to the info as well rather than the guess-logic here it will make running locally a bit less with docker in any scenario where the ports are changed (multiple services for example).

Reply:
Agreed.

Reply:
Although, one of the challenges with docker running locally is that the proxy might not know the port that the user function needs to access it on, since this is specified by a port-forward.